src/Lumi/Components2/ButtonGroup.purs

```diff
@@ -3,28 +3,75 @@ module Lumi.Components2.ButtonGroup where
 import Prelude
 
 import Effect.Unsafe (unsafePerformEffect)
+import Lumi.Components (PropsModifier)
 import Lumi.Components as L
-import Lumi.Styles (toCSS)
-import Lumi.Styles.Button as Styles.Button
+import Lumi.Components.ZIndex (ziButtonGroup)
+import Lumi.Styles (css, nested, px, str, style_, toCSS)
+import Lumi.Styles.Box (_row, box)
 import Lumi.Styles.Theme (useTheme)
 import React.Basic.DOM as R
 import React.Basic.Emotion as E
 import React.Basic.Hooks (JSX)
 import React.Basic.Hooks as React
 
 type ButtonGroupProps
-  = ( joined :: Boolean
+  = ( component :: ButtonGroup
     , content :: Array JSX
     )
 
 buttonGroup :: L.LumiComponent ButtonGroupProps
 buttonGroup =
   unsafePerformEffect do
-    L.lumiComponent "ButtonGroup" { joined: false, content: [] } \props -> React.do
+    L.lumiComponent "ButtonGroup" { component: ButtonGroup, content: [] } \props -> React.do
       theme <- useTheme
       pure
         $ E.element R.div'
             { className: props.className
             , children: props.content
-            , css: theme # toCSS (Styles.Button.buttonGroup props.joined) <> props.css
+            , css: theme # toCSS styles <> props.css
             }
+
+  where
+  styles =
+    box
+      <<< _row
+      <<< style_
+            ( css
+                { label: str "buttonGroup"
+                , "& > *:not(:last-child)":
+                    nested
+                      $ css
+                          { marginRight: px 8
+                          }
+                }
+            )
+
+
+data ButtonGroup = ButtonGroup
+
+type ButtonGroupModifier = forall r. PropsModifier ( component :: ButtonGroup | r )
+
+joined :: ButtonGroupModifier
+joined =
+  style_
+    $ css
+        { label: str "joined"
+        , "& > *:not(:last-child)":
+            nested
+              $ css
+                  { marginRight: px (-1)
+                  , borderTopRightRadius: px 0
+                  , borderBottomRightRadius: px 0
+                  }
+        , "& > *:not(:first-child)":
+            nested
+              $ css
+                  { borderTopLeftRadius: px 0
+                  , borderBottomLeftRadius: px 0
+                  }
+        , "& > *:focus, & > *:hover":
+            nested
+              $ css
+                  { zIndex: px ziButtonGroup
+                  }
+        }
```

Comment:
I like this pattern, thanks @arthurxavierx! These styles feel a lot safer bound to a specific component type.

Reply:
The caveat here is that modifiers of type `forall c. ButtonModifier c` can be applied to any component, so maybe this alternative is safer?
```purescript
type ButtonModifier c = forall r. PropsModifier (buttonComponent :: c | r)
```

And for every component we have `(textComponent :: c | r)`, `(buttonGroupComponent :: c | r)`, etc.

Reply:
Nvm, just saw that `ButtonModifier` constrains the component by requiring some other props 😅

Reply:
Ah, true. I wonder if `PropsModifier` should contain the `(component :: c | props)`.. then it'd be as general as it is now by default so long as `c` is left open, but locked to a specific component as soon as you set `c`, and we wouldn't have the redundant `XYZModifier` types.

Related- I was also wondering if there was a better way to type `loadingContent`, it didn't fit into `ButtonModifier` because of the other prop constraints. It works as is, the inconsistent type  aliases just seemed a little weird.

Reply:
The `XYZModifier` types are also kind of a weird pattern to me. It's just duck typing actually, because we're saying that such and such modifiers apply to any component that has those props (and only to those), and it breaks in the case where a related component has different props (like with `loadingContent`).

I think that what we want is primarily a way of constraining modifiers to certain components in a hierarchy, it seems we don't have to care about the props like that, so maybe another way of doing it could be something like this?
```purescript
type ButtonModifier props = forall c. IsButton c => PropsModifier (component :: c | props)

class IsButton c
instance IsButton Button
instance IsButton LinkButton
```

Not sure how type inference would be affected by that though…

Reply:
Perhaps a phantom type like `PropsModifier Button r` as well? Going to merge this as is for now though, since it doesn't affect usage much.

Reply:
Unfortunately `PropsModifier` cannot have a phantom type parameter because it's a type synonym, so that `PropsModifier Button r` and `PropsModifier Text r` would be the same type (unless that parameter were used in the definition of `PropsModifier`).

Reply:
Ah right